server/src/main/java/org/apache/kafka/server/BrokerFeatures.java

```diff
@@ -0,0 +1,179 @@
+            }
+        });
+        return Features.supportedFeatures(features);
+    }
+
+    public static BrokerFeatures createEmpty() {
+        return new BrokerFeatures(Features.emptySupportedFeatures());
+    }
+
+    /**
+     * Returns true if any of the provided finalized features are incompatible with the provided
+     * supported features.
+     *
+     * @param supportedFeatures The supported features to be compared
+     * @param finalizedFeatures The finalized features to be compared
+     * @return - True if there are any feature incompatibilities found.
+     * - False otherwise.
+     */
+    public static boolean hasIncompatibleFeatures(Features<SupportedVersionRange> supportedFeatures,
+                                                  Map<String, Short> finalizedFeatures) {
+        return !incompatibleFeatures(supportedFeatures, finalizedFeatures, false).isEmpty();
+    }
+
+    /**
+     * Returns the default finalized features that a new Kafka cluster with IBP config >= IBP_2_7_IV0
+     * needs to be bootstrapped with.
+     */
+    public Map<String, Short> defaultFinalizedFeatures() {
+        return supportedFeatures.features().entrySet()
+                .stream()
+                .collect(Collectors.toMap(Map.Entry::getKey,
+                        e -> e.getKey().equals(KRaftVersion.FEATURE_NAME) ? 0 : e.getValue().max()));
+    }
+
+    /**
+     * Returns the set of feature names found to be incompatible.
+     * A feature incompatibility is a version mismatch between the latest feature supported by the
+     * Broker, and a provided finalized feature. This can happen because a provided finalized
+     * feature:
+     * 1) Does not exist in the Broker (i.e. it is unknown to the Broker).
+     * [OR]
+     * 2) Exists but the FinalizedVersionRange does not match with the SupportedVersionRange
+     * of the supported feature.
+     *
+     * @param finalized The finalized features against which incompatibilities need to be checked for.
+     * @return The subset of input features which are incompatible. If the returned object
+     * is empty, it means there were no feature incompatibilities found.
+     */
+    public Map<String, Short> incompatibleFeatures(Map<String, Short> finalized) {
+        return BrokerFeatures.incompatibleFeatures(supportedFeatures, finalized, true);
+    }
+
+    public Features<SupportedVersionRange> supportedFeatures() {
+        return supportedFeatures;
+    }
+
+    private static Map<String, Short> incompatibleFeatures(Features<SupportedVersionRange> supportedFeatures,
+                                                           Map<String, Short> finalizedFeatures,
+                                                           boolean logIncompatibilities) {
+        Map<IncompatibleFeaturesInfo, String> incompatibleFeaturesInfoToErrorReason = 
+                transferToIncompatibleFeaturesInfoToErrorCode(supportedFeatures, finalizedFeatures);
```

Comment:
Could you please inline transferToIncompatibleFeaturesInfoToErrorCode? We can loop through the features and log any incompatible ones. This would simplify the code and provide a slight speed improvement.

Reply:
If use StringBuilder to resolve this errorReason, we should deal with last two ", " so the code will be like 
```java
    private static Map<String, Short> incompatibleFeatures(Features<SupportedVersionRange> supportedFeatures,
                                                           Map<String, Short> finalizedFeatures,
                                                           boolean logIncompatibilities) {
        IncompatibleFeaturesInfosToErrorReasons incompatibleFeaturesInfoToErrorReason = 
                transferToIncompatibleFeaturesInfoToErrorCode(supportedFeatures, finalizedFeatures);
        if (logIncompatibilities && !incompatibleFeaturesInfoToErrorReason.errorReason.isEmpty()) {
            log.warn("Feature incompatibilities seen: {}", incompatibleFeaturesInfoToErrorReason.errorReason);
        }
        return incompatibleFeaturesInfoToErrorReason.incompatibleFeaturesInfo;
    }

    private static IncompatibleFeaturesInfosToErrorReasons transferToIncompatibleFeaturesInfoToErrorCode(
            Features<SupportedVersionRange> supportedFeatures,
            Map<String, Short> finalizedFeatures
    ) {
        Map<String, Short> incompatibleFeaturesInfo = new HashMap<>();
        StringBuilder errorReasons = new StringBuilder();
        finalizedFeatures.forEach((feature, versionLevels) -> {
            SupportedVersionRange supportedVersions = supportedFeatures.get(feature);
            if (supportedVersions == null) {
                incompatibleFeaturesInfo.put(feature, versionLevels);
                errorReasons.append(format("{feature=%s, reason='Unsupported feature'}, ", feature));
            } else if (supportedVersions.isIncompatibleWith(versionLevels)) {
                incompatibleFeaturesInfo.put(feature, versionLevels);
                errorReasons.append(format("{feature=%s, reason='%s is incompatible with %s'}, ", feature, versionLevels, supportedVersions));
            }
        });
        return new IncompatibleFeaturesInfosToErrorReasons(
                incompatibleFeaturesInfo,
                errorReasons.length() == 0 ? "" :
                errorReasons.delete(errorReasons.length() - 2, errorReasons.length()).toString()
        );
    }
    
    private static class IncompatibleFeaturesInfosToErrorReasons {
        final Map<String, Short> incompatibleFeaturesInfos;
        final String errorReason;
        
        IncompatibleFeaturesInfosToErrorReasons(Map<String, Short> incompatibleFeaturesInfos, String errorReason) {
            this.incompatibleFeaturesInfos = incompatibleFeaturesInfos;
            this.errorReason = errorReason;
        }
    }
```
I prefer to use the List to resolve this, It is more readable
```java
    private static Map<String, Short> incompatibleFeatures(Features<SupportedVersionRange> supportedFeatures,
                                                           Map<String, Short> finalizedFeatures,
                                                           boolean logIncompatibilities) {
        IncompatibleFeaturesInfosToErrorReasons incompatibleFeaturesInfoToErrorReason = 
                transferToIncompatibleFeaturesInfoToErrorCode(supportedFeatures, finalizedFeatures);
        if (logIncompatibilities && !incompatibleFeaturesInfoToErrorReason.errorReason.isEmpty()) {
            log.warn("Feature incompatibilities seen: {}", incompatibleFeaturesInfoToErrorReason.errorReason);
        }
        return incompatibleFeaturesInfoToErrorReason.incompatibleFeaturesInfos;
    }

    private static IncompatibleFeaturesInfosToErrorReasons transferToIncompatibleFeaturesInfoToErrorCode(
            Features<SupportedVersionRange> supportedFeatures,
            Map<String, Short> finalizedFeatures
    ) {
        Map<String, Short> incompatibleFeaturesInfo = new HashMap<>();
        List<String> errorReasons = new ArrayList<>();
        finalizedFeatures.forEach((feature, versionLevels) -> {
            SupportedVersionRange supportedVersions = supportedFeatures.get(feature);
            if (supportedVersions == null) {
                incompatibleFeaturesInfo.put(feature, versionLevels);
                errorReasons.add(format("{feature=%s, reason='Unknown feature'}", feature));
            } else if (supportedVersions.isIncompatibleWith(versionLevels)) {
                incompatibleFeaturesInfo.put(feature, versionLevels);
                errorReasons.add(format("{feature=%s, reason='%s is incompatible with %s'}", feature, versionLevels, supportedVersions));
            }
        });
        return new IncompatibleFeaturesInfosToErrorReasons(
                incompatibleFeaturesInfo,
                String.join(", ", errorReasons)
        );
    }
    
    private static class IncompatibleFeaturesInfosToErrorReasons {
        final Map<String, Short> incompatibleFeaturesInfos;
        final String errorReason;
        
        IncompatibleFeaturesInfosToErrorReasons(Map<String, Short> incompatibleFeaturesInfos, String errorReason) {
            this.incompatibleFeaturesInfos = incompatibleFeaturesInfos;
            this.errorReason = errorReason;
        }
    }
```
WDYT?